src/error.rs

```diff
@@ -76,26 +105,20 @@ pub(crate) fn convert_located<T: PartialEq>(datas: Vec<T>) -> Vec<Located<T>> {
     datas.into_iter().map(|d| Located::from_data(d)).collect()
 }
 
-macro_rules! invalid_token {
-    ($location: expr, $($arg:tt)*) => (
-        return Err(SchemeError {category: ErrorType::Lexical, message: format!($($arg)*), location: $location });
-    )
-}
-
 macro_rules! syntax_error {
```

Comment:
``syntax_error!(SyntaxError:: ...)`` and ``logic_error!(LogicError:: ...)`` is complicated,
if ``impl From<Located<SyntaxError>> for SchemeError`` and ``impl From<Located<LogicError>> for SchemeError``, then it is safe to use ``Err(Located<SyntaxError> or Located<LogicError>)?`` when return type is Result<_, SchemeError>

Reply:
It would be more verbose. For example, 
`Err(From::from(Located<SyntaxError>::new(SyntaxError::UnexpectedEnd), location)))` is more complitcated.than `syntax_error!(SyntaxError::UnexpectedEnd, location)`

You cannot directly return an unconverted error, but can only emit the conversion with `?`.

Reply:
It's just ``Err(SyntaxError::UnexpectedEnd.locate(location).into())``
or ``Err(SyntaxError::UnexpectedEnd.locate(location))?``, not need to repeat "syntax" twice
could use a ``error!`` macro both for syntax and logic, like ``error!(SyntaxError::UnexpectedEnd, location)`` and ``error!(LogicError::SomeLogicError, location)``

Reply:
There is no need to introduce `Located<SyntaxError>`, and `impl From<Located<SyntaxError>>` is conflict with `thiserror`

Reply:
Look like you already ``impl From<Located<SyntaxError>> for SchemeError``?
https://github.com/Danielmelody/Ruschm/blob/2a89bbbe69291d2caa8977cdc2d40b777fcde59a/src/parser/error.rs#L46

Reply:
Yes, I forget to delete it. Updated.
T think replace `refactor xxx_error` with `error` by using `ErrorData::from` is ok.